modules/n1ql/pages/n1ql-language-reference/recursive.adoc

```diff
@@ -0,0 +1,315 @@
+= Recursive WITH Clause
+:description: Use the RECURSIVE keyword to enable recursive referencing in WITH clause common table expressions.
+:imagesdir: ../../assets/images
+:page-topic-type: reference
+
+[abstract]
+{description}
+
+[#purpose]
+== Purpose
+
+The common table expressions (CTEs) created by the WITH clause simplify complex queries and create temporary result sets which can be used as data sources or as expressions for queries.
+
+The SELECT definition for CTEs is split by the UNION/UNION ALL operators.
+
+* The left arm of UNION/UNION ALL contains an anchor clause, which is non-recursive. 
+The anchor clause produces an initial set of documents as an intermediate result set for the CTE keyspace.
+
+* The right arm of UNION/UNION ALL contains the recursive clause.
+The recursive clause produces a new set of documents.
+
+The recursive clause is executed again, with the CTE keyspace now referring to the result produced by the initial recursive clause.
+This process repeats until it results in an empty intermediate result set.
+The intermediate result sets are appended to produce a final output.
+
+[#syntax]
+== Syntax
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause]
+----
+
+image::n1ql-language-reference/recursive-with-clause.png["Syntax diagram", align=left]
+
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-union]
+----
+
+image::n1ql-language-reference/recursive-with-clause-union.png["Syntax diagram UNION/UNION ALL", align=left]
+
+[#arguments]
+== Arguments
+
+The following arguments are optional.
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-options]
+----
+
+image::n1ql-language-reference/recursive-with-clause-options.png["Argument diagram", align=left]
+
+options:: Object that signals an early exit.
+* `levels`: An exit after level `N`. 
+This object assumes the Anchor is at level 0.
+* `documents`: An exit after accumulating `N` documents. 
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-cycle]
+----
+
+image::n1ql-language-reference/recursive-with-clause-cycle.png["Argument diagram", align=left]
+
+cycle:: `CYCLE <fieldnames_list> RESTRICT`. 
+List of field names in which you want to perform cycle detection.
```

Comment:
Use individual 3rd level headings for each subclause in the syntax, rather than a catch-all "Arguments" section. Do the CYCLE clause first, because it comes first in the syntax. Use a table to describe the schema for the OPTIONS object.

Note also that we don't need to include any more than minimal alt text for the syntax diagrams; the EBNF text _is_ the alt text for the syntax diagram.

```suggestion
[[cycle-clause]]
=== CYCLE Clause

[source,ebnf]
----
include::partial$grammar/dql.ebnf[tag=cycle-clause]
----

image::n1ql-language-reference/cycle-clause.png["Syntax diagram", align=left]

The optional CYCLE clause provides one method for avoiding infinite recursions.
It enables you to specify one or more fields whose values are likely to repeat.

expr::
(Required) An identifier representing a field.

[[options-clause]]
=== OPTIONS Clause
[source,ebnf]
----
include::partial$grammar/dql.ebnf[tag=options-clause]
----

image::n1ql-language-reference/options-clause.png["Syntax diagram", align=left]

The optional OPTIONS clause provides another method for avoiding infinite recursions.
It enables you to specify that the recursion should exit after a specified level, or after accumulating a specified number of documents.

expr::
(Required) An object with the following properties.

[options="header", cols="~a,~a,~a"]
|===
|Name|Description|Schema

|**levels** +
__optional__
|Recursion should exit after reaching this level.
This assumes the anchor is at level 0.
|Integer

|**documents** +
__optional__
|Recursion should exit after accumulating this many documents.
|Integer
|===
```